File: filemaid/cleaners/cargo.py
"""cargo cache cleanup."""
import os
import shutil
import subprocess
from pathlib import Path

from filemaid.cleaners._result import CleanupResult
from filemaid.cleaners._size import dir_size, humanize


def _cargo_cache_dir() -> Path:
    return Path(os.environ.get("CARGO_HOME", Path.home() / ".cargo")) / "registry" / "cache"


def can_run() -> bool:
    return shutil.which("cargo") is not None and shutil.which("cargo-cache") is not None


def run(dry_run: bool, config: dict) -> CleanupResult:
    cache_dir = _cargo_cache_dir()
    before = dir_size(cache_dir) if cache_dir.exists() else 0
    if dry_run:
        return CleanupResult(
            name="cargo",
            status="dry-run",
            saved=before,
            saved_human=humanize(before),
            detail="would autoclean cache",
        )
    try:
        result = subprocess.run(
            ["cargo", "cache", "--autoclean"],
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
            timeout=300,
        )
        after = dir_size(cache_dir) if cache_dir.exists() else 0
        saved = max(0, before - after)
        output = result.stdout.strip()
        return CleanupResult(
            name="cargo",
            status="ok",
            saved=saved,
            saved_human=humanize(saved),
            detail=f"cache autocleaned\n{output}" if output else "cache autocleaned",
        )
    except Exception as exc:
        return CleanupResult(
            name="cargo",
            status="failed",
            detail=str(exc),
        )

File: internal/cleaners/cargo.go
package cleaners

import (
	"os"
	"path/filepath"
	"strings"

	"github.com/logicminds/filemaid/internal/config"
)

var cargoCleaner = Cleaner{
	Name:   "cargo",
	CanRun: cargoCanRun,
	Run:    cargoRun,
}

func cargoCanRun() bool {
	_, err := lookPath("cargo")
	if err != nil {
		return false
	}
	_, err = lookPath("cargo-cache")
	return err == nil
}

func cargoCacheDir() string {
	cargoHome := os.Getenv("CARGO_HOME")
	if cargoHome == "" {
		cargoHome = filepath.Join(home(), ".cargo")
	}
	return filepath.Join(cargoHome, "registry", "cache")
}

func cargoRun(dryRun bool, cfg *config.Config) CleanupResult {
	cacheDir := cargoCacheDir()
	before := dirSizeIfExists(cacheDir)

	if dryRun {
		saved := before
		return CleanupResult{
			Name:       "cargo",
			Status:     "dry-run",
			Saved:      &saved,
			SavedHuman: Humanize(before),
			Detail:     "would autoclean cache",
		}
	}

	out, err := runner.Run("cargo", "cache", "--autoclean")
	if err != nil {
		return CleanupResult{
			Name:   "cargo",
			Status: "failed",
			Detail: err.Error(),
		}
	}

	after := dirSizeIfExists(cacheDir)
	saved := before - after
	if saved < 0 {
		saved = 0
	}
	output := strings.TrimSpace(string(out))
	detail := "cache autocleaned"
	if output != "" {
		detail = "cache autocleaned\n" + output
	}

	return CleanupResult{
		Name:       "cargo",
		Status:     "ok",
		Saved:      &saved,
		SavedHuman: Humanize(saved),
		Detail:     detail,
	}
}
